### ISSUE TITLE
Some thoughts on IFQL

### ISSUE THREAD

**Issue by DanCech** [influxdata/ifql#144]
I'm very excited about the possibilities for IFQL after seeing @pauldix presentation at Influx/Days!

After having some time to mull things over, I wanted to share some of my thoughts on areas that I think might be able to be made more intuitive for users.

The biggest thing that I find confusing is the co-mingling of orm-style functions that build the spec for collecting data from the backend with the pipeline of processing functions that operate on the data once it has been retrieved.  This may be more of a documentation/education issue than anything else, but it seems like it should be clearer that there are really 2 separate things going on, and the docs need to communicate clearly any associated restrictions and/or semantics around using for example multiple calls to `.range()` or `.where()`.

Conceptually I find it easy to think of the initial `select/where/range` (or `from/where/range`) as building the actual data query, then the remaining functions in the pipeline as performing processing on the results, though I haven't looked at the code to know whether that's accurate or if for example the first `.window()` gets pushed down also.

The way that the aggregation function is specified for `window` is the next really confusing thing, a call like `.window(...).min().max()` is very confusing for a user to parse.  It seems like it should really be passed as a parameter, especially if you think of `window` like a super-`map` that runs a callback against each slice of the series to calculate the new value.  If you think of each window as a tiny series result in its own right and a chain of filter functions as a higher-order "function" that returns a transformation function that is then executed on the series, it makes sense to support a syntax like `.window(every:10m, reduce:min())` or later on something like `.window(every:10m, reduce:percentile(percentile:90))`.

If you continue that way of thinking about the processing part of the chain as applying the result of a higher-order function to the timeseries you could arrive at a potential solution for the first issue with something like `select(db:"telegraf").where(exp:{"_measurement" == "cpu" AND "_field" == "usage_user"}).range(start:-170h).apply(sum())`.  That syntax seems awkward too, but it does make it clearer that there is a separation between the 2 phases so it might be a useful thought experiment.

In a similar vein, the use of `.where()` as both a selector and a filter is confusing, a query with multiple `where` calls makes it seem as though the whole thing is defining a query orm-style that's going to be executed at the end of the chain, and as a user I'd expect the second call to be added to the first one as an additional selector, not used in the processing pipeline as a filter.  The simple fix for that is to have a separate function (`filter`?) to use in that scenario.

Finally (for now) the where clause syntax.  Dropping the `{}` seems like it would be an improvement, though I do worry about clarity if `where` ends up supporting multiple parameters.  I do like the idea of being able to chain things together with `AND` & `OR`, though I don't see any mention in the docs of syntax like `("x" == "y" OR x == "z") AND (...)`.  Also unsure about the need for quotes around the tag/field names.  Explicit support for multiple where conditions (`AND`ed together in the final query) would be a good thing as mentioned above.  I'd also suggest that the existing `$` syntax be replaced with `_value`.

Anyway, I hope some of that might be helpful and am more than happy to discuss in more detail.

**Comment by nathanielc** [comment]
Hey @DanCech thanks for the feedback!

I like to think of the query as not really being a mix of orm-style function and pipeline functions, but that all functions are pipeline functions.

For example this is how I read the query (using the newly proposed changes):

```javascript
var cpu = from(db:"telegraf") // request all data from the database "telegraf"
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total") // filter data by the expression
    .range(start:1-h) // filter the data to be within the time bounds
    .sum() // compute the sum for each series
```

Naturally that is not what actually happens. The planning step recognizes that you can push the first operations down into a single read operation on the storage layer. But the operations are still consistent in meaning. A `filter` operation filters the data that passes through that step of the pipeline, no matter if its at the beginning or in the middle of a query. Is just so happens that the engine is smart enough to optimize the initial filters in order to reduce the amount of data transferred. Does that mental model help? Or is it more confusing? 


On thing that is missing from the docs at the moment is a good discussion on the data model that IFQL uses, so the reader is left guessing a bit. In short tables are passed from step to step in the query. Each operation has a well defined behavior. Operations are functions that accepts a tables as input and produces tables as output. This way each operation has an easily understood behavior. The user should not feel the need to distinguish between operations that can be pushed down to the storage tier or operations that run in the query engine. The behavior is the same regardless. 


Thanks for the notes on adding docs on `AND` and `OR` for expressions. They are supported.

**Comment by DanCech** [comment]
Interesting, I can see what you're getting at in terms of the user not having to think about what is a selector vs a filter and keeping the door open for pushing more of the pipeline down to the individual backends where possible (I've had some similar thoughts about graphite clustering).

If you want to look at things that way, then I'm a fan of using `filter` rather than `where`, but I'd also be concerned that by hiding too many of the implementation details behind the "everything's a filter" abstraction you're at greater risk of the user writing inefficient queries without realizing what they're doing.  I'm assuming the planner is smart enough to push multiple `.filter()` calls down into the storage layer if possible?

The current syntax for `.window()` is also a big thing that concerns me (see previous post), since the same function are used to perform 2 very different functions (aggregating across windows and aggregating across the entire series).

**Comment by r3cgm** [comment]
Just a small thought, do we really need to do a double equals for an equality comparison, as in:
.filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
?

Can we get away with a single equals:
.filter(exp: "_measurement" = "cpu" AND "_field" = "usage_idle" AND "cpu" = "cpu-total")
?

... or will it actually be possible to use equals as an assignment operator inside a .filter() or .where() clause so we need to disambiguate?  I'm just thinking given how frequently these clauses are going to be typed that simplifying syntax would be welcome.  This is along the same lines as the suggestion that we be able to use unquoted bare fieldnames.

Thanks for any consideration.

**Comment by DanCech** [comment]
Yup, I had the same thought and question :)

**Comment by nathanielc** [comment]
Good question, its likely we will support assignment in expressions. Have a look at #149 which explores making it more clear that the argument to filter is a function not just a boolean expression. In that context its likely to be normal to assign intermediate values.

As for the window function I have some thoughts, quoting @DanCech:

> The current syntax for .window() is also a big thing that concerns me (see previous post), since the same function are used to perform 2 very different functions (aggregating across windows and aggregating across the entire series).

I see the confusion here as I was initially confused by the meaning of two chained aggregate functions. What follows is how I understand the flow, its not meant to say this is how it has to be but rather to continue our discussion. 

**TL;DR**

The aggregate functions are represented as the same function in IFQL because they are the same function, both in concept and implementation. An aggregate function operates on bounded data, independent of how the bounds were created.

Further explanation with examples:

A query can have a time boundary, this is set using the range function.  For example, the following query performs the `sum` aggregation over the time bounds defined by the `range` function.

```
var cpu = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:1-h)
    .sum()
```

Alternatively a query does not need a `range` function. This means the query runs for all time, (think of it like a continuous query or Kapacitor task). In this case we can add a window function to create discrete time bounds on the stream of unbounded data.

```
var cpu = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .window(every:10m) // introduce new time bounds at 10m intervals
    .count() // aggregate across the window time bounds, repeatedly forever.
```

It also a valid query to combine both of these queries.

```
var cpu = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:-1h)
    .window(every:10m) // introduce new time bounds at 10m intervals
    .count() // aggregate across the window time bounds, but only 6 times since an hour has 6, 10m intervals.
    .sum() // aggregate across the range time bounds, which are the same time bounds as without windowing
```

The concepts of time bounds and aggregates are independent of how the time bounds where set.
Meaning that an aggregate like `sum` is simply defined as computing the sum for a bounded set of data in time. How those bounds are set are not important to the `sum` aggregate. The bounds could be from the window function or the range function, or possibly some other function in the future. 

The confusing part to me(when I first read it) was its not clear that after aggregating with a window, that the time bounds are set back to the bounds from the range function. In hind sight it makes sense because the input data (to the window) was bounded, so the output data (from the window+aggregate) would also be bounded. 

In short aggregate functions, always return a single record for a set of bounded data, independent of how the bounds were placed on the data. This means that the there is no special meaning to an aggregate after a window function as opposed to after a range function. This is why they are represented as being the same in the language because their behavior is the same.

Thoughts?  Also @pauldix do you want to chime in here on how you are thinking about multiple chained aggregates?

**Comment by DanCech** [comment]
I must confess that the explanation of window above doesn't make sense to me, what is the result of the following query?

```
var cpu = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:-1h)
    .window(every:10m)
```

**Comment by nathanielc** [comment]
The client will receive 6 blocks of data for the cpu=cpu-total series. Each block of data will only contain data for a single time interval. Specifically the time bounds of each block will be 0-10, 10-20, ..., 50-60. 

Returning windowed data to a client, specially a client that is going to graph the data doesn't make a lot of sense, as the client will just display them all in a series anyways. But other types of clients will want the data returned in its windowed form, specially when the windowing operation was more complicated than a simple fixed interval.

**Comment by DanCech** [comment]
Cool, that makes sense.

I still have an issue with the use of the aggregator as a chained function after the window operation though.  If I understand it correctly, you're starting with a result set that has a single "window" encompassing the whole range, then using `.window()` to break it up into multiple smaller windows, so the next aggregator then takes those and turns them back into a single large "window" containing the aggregated points from each window? Then a second aggregator will then aggregate the points in that "window" down to a single point. Is that right?

This still seems like it would be more intuitive if the window function was able to accept a chain of functions to apply to the points in each window, not only because it follows the existing map/reduce pattern most people are familiar with, but because it makes it completely obvious which functions are applied to the entire result set (the ones in the main pipeline) and which are applied to each window (the ones passed to `.window()`) without having to think about whether the result set has been windowed or not.

So this:
```
var cpu = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:-1h)
    .window(every:10m) // introduce new time bounds at 10m intervals
    .count() // aggregate across the window time bounds, but only 6 times since an hour has 6, 10m intervals.
    .sum() // aggregate across the range time bounds, which are the same time bounds as without windowing
```

Would become something like:
```
var cpu = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:-1h)
    .window(
        every:10m, // introduce new time bounds at 10m intervals
        reduce: count() // aggregate across the window time bounds, but only 6 times since an hour has 6, 10m intervals.
    )
    .sum() // aggregate across the range time bounds, which are the same time bounds as without windowing
```

Anyway, I'm sure users will get accustomed to it regardless, just wanted to lend some input from the vantage point of someone coming in with some experience with TSDBs but not being familiar with IFQL and the things that jumped out at me as being strange/confusing.

**Comment by nathanielc** [comment]
@DanCech Thanks your explanation makes sense and we appreciate the feedback.

**Comment by pauldix** [comment]
@nathanielc's explanation of how I think about `window` matches what I was thinking. One other thing to note is that the behavior can be more complex than simply producing boundaries along which aggregates can be computed. For example, if you do:

```javascript
var cpu = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:-1h)
    .window(every:5m ,period:10m) // every 5m produce last 10m of data
```

That would actually duplicate data and return it back. Although the client would then have to deal with it because you'd get tables of data that weren't in strict time order. I think this example makes more sense in the context of a streaming query. Like, product a tick every 5 minutes that is a summary of the last 10 minutes you've seen.

Originally, I had the aggregate function as a required argument to the window function. But through some back and forth with the team people thought it was weird that it was being treated differently. Unfortunately, it wasn't captured on the PR, but I remember talking to @rbetts about it specifically.

I'd be open to passing it as a required argument to window. Like: 

```javascript
```

**Comment by pauldix** [comment]
Ok, added some comments to #149. I think if we get that resolved, the only real question for `window` is if we should be passing functions to it explicitly, which I'm in favor of. Unless there are compelling reasons to not have the function passed in. So something like:

```javascript
from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:-1h)
    .window(every:1m, period: 10m, apply: function(series) {series.max()})
```

**Comment by nathanielc** [comment]
@pauldix What about use cases where you want to apply multiple functions to the windowed data.

Using the current syntax this is straightforward.

```javascript
var data = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:-1h)
    .window(every:1m, period: 10m)

data.min()
data.max()
data.mean()
```

What about use cases where the client wants the data returned in its windowed form?

In addition to windowing the data based on fixed intervals  it will be possible to define windows based on more complex concepts. One example is windowing based on user sessions, i.e points from a single user get added to a window until there is more than 1h of idle time between points. The idle time defines the window boundaries. As a result the window boundaries are different for each series and not regular lengths. In these cases it could make sense to return the raw window data to the client. This way a client could graph the different sessions for a user (possibly using time shift to overlay them).

In short while the common case is to perform an aggregation or a selection on a window that is not always the case. As a result it doesn't make sense to me to couple a function to a window operation.

**Comment by pauldix** [comment]
@nathanielc right, forgot about the multiple functions issue. Now I remember it. Wondering if there's another way to make this less confusing to the user. Or maybe it's just an education thing where we need to show multiple examples of `window` with following aggregation calls and what it actually returns.

**Comment by DanCech** [comment]
@nathanielc in your example wouldn't it be trivial to rewrite it as:
```javascript
var data = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:-1h)

data.window(every:1m, period: 10m, apply: function(series) {series.min()})
data.window(every:1m, period: 10m, apply: function(series) {series.max()})
data.window(every:1m, period: 10m, apply: function(series) {series.mean()})
```
or if you want a more compact syntax where min/max/mean are higher-order functions:
```javascript
data.window(every:1m, period: 10m, apply: min())
data.window(every:1m, period: 10m, apply: max())
data.window(every:1m, period: 10m, apply: mean())
```

**Comment by akansal1** [comment]
I use **pandas** a lot in my work, and also, deal with time-series data. Pandas API has aggregate that allow you to evaluate multiple functions including custom function for the sampling period chosen. Keeping that in mind recommendation of @nathanielc  can be generalized as:

```go
var data = from(db:"telegraf") 
    .filter(exp: "_measurement" == "cpu" AND "_field" == "usage_idle" AND "cpu" == "cpu-total")
    .range(start:-1h)
    .window(every:1m, period: 10m)

data.apply('min' : min(), 'max':max(),'custom':custom())
```
Also, it will be great if, you allow person to export data in binary format like **Apache Arrow**, that can enable developing further solutions over this, easily

**Comment by nathanielc** [comment]
@akansal1 Thanks for the feedback.

>Also, it will be great if, you allow person to export data in binary format like parquet using Apache Arrow in-memory columnar memory format, that can enable developing further solutions over this, easily

We have already been discussion this internally, glad so see you think its a good idea as well.

**Comment by nathanielc** [comment]
Thanks all for the feedback! We have been discussing this internally for quite sometime now and have arrived at a final design. The window function will be independent of the various aggregate functions. 
There are two main reasons for this design:

 * Functions are the composable unit of query operations, requiring that the window function accept an aggregate of selector operation makes the window function less composable.
* The IFQL data model is different from other systems that have merged the concepts of windowing with aggregation. Specifically the data model can express raw windowed data and we want to expose that to end users. We have made several changes to the data model in order to make this a bit more clear, see https://github.com/influxdata/ifql/pull/291

We are in the process of publishing a final specification of IFQL its functions and the data model. That specification will address directly how windowing works.
